apis/python/src/tiledbsoma/_common_nd_array.py

```diff
@@ -73,70 +84,92 @@ def reshape(self, shape: Tuple[int, ...]) -> None:
-
-    # check on shape
-    if len(shape) == 0 or any(e <= 0 for e in shape):
-        raise ValueError("SOMA NDArray shape must be non-zero length tuple of ints > 0")
-
-    if not pa.types.is_primitive(type):
-        raise TypeError(
-            f"Unsupported type {type} --"
-            " SOMA NDArrays only support primtive Arrow types"
-        )
-
-    dims = []
-    for n, e in enumerate(shape):
-        dim_name = f"soma_dim_{n}"
-        dim = tiledb.Dim(
-            name=dim_name,
-            domain=(0, e - 1),
-            tile=min(e, create_options.dim_tile(dim_name, 2048)),
-            dtype=np.int64,
-            filters=create_options.dim_filters(
-                dim_name,
-                [
-                    dict(
-                        _type="ZstdFilter",
-                        level=create_options.sparse_nd_array_dim_zstd_level(),
-                    )
-                ],
-            ),
-        )
-        dims.append(dim)
-    dom = tiledb.Domain(dims, ctx=context.tiledb_ctx)
-
-    attrs = [
-        tiledb.Attr(
-            name="soma_data",
-            dtype=_arrow_types.tiledb_type_from_arrow_type(type),
-            filters=create_options.attr_filters("soma_data", ["ZstdFilter"]),
+    @classmethod
+    def _build_tiledb_schema(
+        cls,
+        type: pa.DataType,
+        shape: Sequence[Union[int, None]],
+        create_options: TileDBCreateOptions,
+        context: SOMATileDBContext,
+        *,
+        is_sparse: bool,
+    ) -> tiledb.ArraySchema:
+        _util.check_type("type", type, (pa.DataType,))
+
+        if not pa.types.is_primitive(type):
+            raise TypeError(
+                f"Unsupported type {type} --"
+                " SOMA NDArrays only support primtive Arrow types"
+            )
+
+        if not shape:
+            raise ValueError("SOMA NDArrays must have a nonzero number of dimensions")
+
+        # check on shape
+        if is_sparse:
+            if any(e is not None and e <= 0 for e in shape):
+                raise ValueError(
+                    "SOMASparseNDArray shape must be non-zero length tuple of positive ints or Nones"
+                )
+        else:
+            if any(e is None or e <= 0 for e in shape):
+                raise ValueError(
+                    "SOMADenseNDArray shape must be non-zero length tuple of positive ints"
+                )
+
+        dims = []
+        for n, e in enumerate(shape):
+            dim_name = f"soma_dim_{n}"
+
+            slot_capacity = 2**63 if e is None else e
+            extent = min(slot_capacity, create_options.dim_tile(dim_name, 2048))
+            # TileDB requires that each signed-64-bit-int domain slot, rounded up to
+            # a multiple of the tile extent in that slot, be representable as a
+            # signed 64-bit int. So if the tile extent is 999, say, this will exceed
+            # 2**63 - 1.
+            if e is None:
+                slot_capacity -= extent
```

Comment:
This should probably still be separated out as in [my previous suggestion](https://github.com/single-cell-data/TileDB-SOMA/pull/952#discussion_r1108916830), so that this doesn’t have to special case sparse vs. not sparse, and the "infinite capacity" case is easier to follow.

Reply:
@thetorpedodog something is wrong -- i very much split these up late last night per your (excellent) suggestions

let me check for a merge fail or somesuch b/c i _did_ do this ... 😬

Reply:
[such things](https://github.com/single-cell-data/TileDB-SOMA/pull/964) befall us all from time to time

Reply:
looking at https://github.com/single-cell-data/TileDB-SOMA/commit/1a25808c8b78951be59c7a94423dfd67d70c26b5 ...

Reply:
@thetorpedodog https://github.com/single-cell-data/TileDB-SOMA/pull/952/commits/e0e31e7ed54d15eb88726265d50fa49aa3352caf -- sorry for the confusion!